ecr/resolver.go

```diff
@@ -241,49 +247,43 @@ type manifestProbe struct {
 	Manifests []json.RawMessage `json:"manifests,omitempty"`
 }
 
-// TODO: add error to signal unparsable and unhandled manifest types.
-func parseImageManifestMediaType(ctx context.Context, body string) string {
+func parseImageManifestMediaType(ctx context.Context, body string) (string, error) {
 	// The unsigned variant of Docker v2 Schema 1 manifest mediaType.
 	const mediaTypeDockerSchema1ManifestUnsigned = "application/vnd.docker.distribution.manifest.v1+json"
 
-	// The type used as a fallback when parsing is not possible.
-	const unparsedMediaType = images.MediaTypeDockerSchema2Manifest
-
 	var manifest manifestProbe
 	err := json.Unmarshal([]byte(body), &manifest)
 	if err != nil {
-		log.G(ctx).WithField("manifest", body).
-			WithError(err).Warn("ecr.resolver.resolve: could not parse manifest")
-		return unparsedMediaType
+		return "", errors.Wrapf(ErrInvalidManifest, "failed to unmarshall %q as a manifest", body)
 	}
 
 	switch manifest.SchemaVersion {
 	case 2:
 		// Defer to the manifest declared type.
 		if manifest.MediaType != "" {
-			return manifest.MediaType
+			return manifest.MediaType, nil
 		}
 		// Is a manifest list.
 		if len(manifest.Manifests) > 0 {
-			return images.MediaTypeDockerSchema2ManifestList
+			return images.MediaTypeDockerSchema2ManifestList, nil
 		}
 		// Is a single image manifest.
-		return images.MediaTypeDockerSchema2Manifest
+		return images.MediaTypeDockerSchema2Manifest, nil
 
 	case 1:
 		// Defer to the manifest declared type.
 		if manifest.MediaType != "" {
-			return manifest.MediaType
+			return manifest.MediaType, nil
 		}
 		// Is Signed Docker Schema 1 manifest.
 		if len(manifest.Signatures) > 0 {
-			return images.MediaTypeDockerSchema1Manifest
+			return images.MediaTypeDockerSchema1Manifest, nil
 		}
 		// Is Unsigned Docker Schema 1 manifest.
-		return mediaTypeDockerSchema1ManifestUnsigned
+		return mediaTypeDockerSchema1ManifestUnsigned, nil
+	default:
+		return "", errors.Wrapf(ErrInvalidManifest, "invalid manifest: %+v", manifest)
```

Comment:
This should say something along the lines of `"unsupported schema version %d"` rather than appending the full manifest blob. The debug log level will already have included the manifest body earlier on and I'd prefer to see terse error strings used where possible.

Reply:
I see. Fixed in the last revision.